src/r1cs/proof.rs

```diff
@@ -56,3 +62,135 @@ pub struct R1CSProof {
     /// Proof data for the inner-product argument.
     pub(super) ipp_proof: InnerProductProof,
 }
+
+impl R1CSProof {
+    /// Serializes the proof into a byte array of \\(16 + 2k\\) 32-byte elements,
+    /// where \\(k=\lceil \log_2(n) \rceil\\) and \\(n\\) is the number of multiplication gates.
+    ///
+    /// # Layout
+    ///
+    /// The layout of the r1cs proof encoding is:
+    ///
+    /// * eleven compressed Ristretto points \\(A_{I1},A_{O1},S_1,A_{I2},A_{O2},S_2,T_1,...,T_6\\),
+    /// * three scalars \\(t_x, \tilde{t}_x, \tilde{e}\\),
+    /// * \\(k\\) pairs of compressed Ristretto points \\(L_0,R_0\dots,L_{k-1},R_{k-1}\\),
+    /// * two scalars \\(a, b\\).
+    pub fn to_bytes(&self) -> Vec<u8> {
+        let mut buf = Vec::with_capacity(self.serialized_size());
+        buf.extend_from_slice(self.A_I1.as_bytes());
+        buf.extend_from_slice(self.A_O1.as_bytes());
+        buf.extend_from_slice(self.S1.as_bytes());
+        buf.extend_from_slice(self.A_I2.as_bytes());
+        buf.extend_from_slice(self.A_O2.as_bytes());
+        buf.extend_from_slice(self.S2.as_bytes());
+        buf.extend_from_slice(self.T_1.as_bytes());
+        buf.extend_from_slice(self.T_3.as_bytes());
+        buf.extend_from_slice(self.T_4.as_bytes());
+        buf.extend_from_slice(self.T_5.as_bytes());
+        buf.extend_from_slice(self.T_6.as_bytes());
+        buf.extend_from_slice(self.t_x.as_bytes());
+        buf.extend_from_slice(self.t_x_blinding.as_bytes());
+        buf.extend_from_slice(self.e_blinding.as_bytes());
+        // XXX this costs an extra alloc
+        buf.extend_from_slice(self.ipp_proof.to_bytes().as_slice());
+        buf
+    }
+
+    /// Returns the size in bytes required to serialize the `R1CSProof`.
+    pub fn serialized_size(&self) -> usize {
+        // 14 elements + the ipp
+        14 * 32 + self.ipp_proof.serialized_size()
+    }
+
+    /// Deserializes the proof from a byte slice.
+    ///
+    /// Returns an error if the byte slice cannot be parsed into a `R1CSProof`.
+    pub fn from_bytes(mut slice: &[u8]) -> Result<R1CSProof, R1CSError> {
```

Comment:
This has the same trailing-bytes issue as RangeProof, but we can fix both in the follow-up PR. Issue filed here: https://github.com/dalek-cryptography/bulletproofs/issues/255